core/src/main/scala/kafka/server/ReplicaAlterLogDirsThread.scala

```diff
@@ -96,57 +97,55 @@ class ReplicaAlterLogDirsThread(name: String,
   }
 
   override def removePartitions(topicPartitions: Set[TopicPartition]): Map[TopicPartition, PartitionFetchState] = {
-    // Schedule assignment request to revert any queued request before cancelling
-    for {
-      topicPartition <- topicPartitions
-      partitionState <- partitionAssignmentRequestState(topicPartition)
-      if partitionState == QUEUED
-      partition = replicaMgr.getPartitionOrException(topicPartition)
-      topicId <- partition.topicId
-      directoryId <- partition.logDirectoryId()
-      topicIdPartition = new TopicIdPartition(topicId, topicPartition.partition())
-    } directoryEventHandler.handleAssignment(topicIdPartition, directoryId, () => ())
+    for (topicPartition <- topicPartitions) {
+      // Revert any reassignments for partitions that did not complete the future replica promotion
+      val PromotionState(reassignmentState, topicId, originalDir) = this.promotionState(topicPartition)
+      if (reassignmentState.inconsistentMetadata) {
+        directoryEventHandler.handleAssignment(new TopicIdPartition(topicId, topicPartition.partition()), originalDir, () => ())
```

Comment:
From [KIP-858](https://cwiki.apache.org/confluence/display/KAFKA/KIP-858%3A+Handle+JBOD+broker+disk+failure+in+KRaft#KIP858:HandleJBODbrokerdiskfailureinKRaft-Intra-brokerreplicamovement):

> In this exceptional case, the broker issues a AssignReplicasToDirs RPC to the Controller to assignment the replica to UUID.LOST_DIR - this lets the Controller know that it needs to update leadership and ISR for this partition too.

But it looks like we only revert the assignment back to the source DIR. Questions:
1. Will the controller do leader election for this partition in this case?
2. If so, should we update KIP? If not, should we follow the KIP?
3. We might need to add tests to make sure leader is re-elected.

Reply:
Good questions.

1. Yes. See `ReplicationController.handleAssignReplicasToDirs`. If a replica is assigned to a directory that is not an online directory in the broker, leadership and ISR is updated.
2. Yes, I'll update the KIP. At the time, I thought it made more sense to use `UUID.LOST_DIR`, but in practice it's a more general solution to just revert the assignment.
3. I've started but haven't yet been able to come up with an integration test for this. It requires blocking the process of replica fetching and assginment processing and injecting failures at very specific times. Let me know if you think it's critical to have an integration test for this, and I'll give it another go.

Reply:
For (3), had a quick look, indeed it's not an easy test. 
Could we, instead, test the logic in `ReplicationControlManager`? That is, add a test in `ReplicationControlManagerTest`, to make sure after receiving a reassignment request ->  log failure request -> reverted reassignment request, then, it should be able to elect another leader... etc?